zipkin/src/main/java/com/linecorp/armeria/server/tracing/TracingService.java

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright 2019 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.server.tracing;
+
+import static com.linecorp.armeria.common.tracing.RequestContextCurrentTraceContext.ensureScopeUsesRequestContext;
+
+import java.util.function.Function;
+
+import com.linecorp.armeria.common.HttpHeaders;
+import com.linecorp.armeria.common.HttpRequest;
+import com.linecorp.armeria.common.HttpResponse;
+import com.linecorp.armeria.common.logging.RequestLogAvailability;
+import com.linecorp.armeria.common.tracing.RequestContextCurrentTraceContext;
+import com.linecorp.armeria.internal.tracing.AsciiStringKeyFactory;
+import com.linecorp.armeria.internal.tracing.SpanContextUtil;
+import com.linecorp.armeria.internal.tracing.SpanTags;
+import com.linecorp.armeria.server.Service;
+import com.linecorp.armeria.server.ServiceRequestContext;
+import com.linecorp.armeria.server.SimpleDecoratingService;
+
+import brave.Span;
+import brave.Span.Kind;
+import brave.Tracer;
+import brave.Tracer.SpanInScope;
+import brave.Tracing;
+import brave.propagation.TraceContext;
+import brave.propagation.TraceContextOrSamplingFlags;
+
+/**
+ * Decorates a {@link Service} to trace inbound {@link HttpRequest}s using
+ * <a href="http://zipkin.io/">Zipkin</a>.
+ *
+ * <p>This decorator retrieves trace data from HTTP headers. The specifications of header names and its values
+ * correspond to <a href="http://zipkin.io/">Zipkin</a>.
+ */
+public final class TracingService extends SimpleDecoratingService<HttpRequest, HttpResponse> {
+
+    /**
+     * Returns a new builder.
+     */
+    public static TracingServiceBuilder builder() {
+        return new TracingServiceBuilder();
+    }
+
+    /**
+     * Creates a new tracing {@link Service} decorator using the specified {@link Tracing} instance.
+     */
+    public static Function<Service<HttpRequest, HttpResponse>, TracingService> decorator(Tracing tracing) {
+        ensureScopeUsesRequestContext(tracing);
```

Comment:
How about adding `requireNonNull()` on `tracing` in `ensureScopeUsesRequestContext()`?